# Formateo de Strings

El formateo de strings nos permite injectar objetos dentro de un string en vez de estar sumandolos o separandolos por coma en el print.

Veamos un ejemplo:

In [4]:
# supongamos que tenemos 
nombre = 'juan'
ciudad = 'cordoba'

print('Su nombres es ' + nombre + ' y vive en la ciudad de ' + ciudad)

Su nombres es juan y vive en la ciudad de cordoba


In [5]:
print('Su nombres es ' , nombre , ' y vive en la ciudad de ' , ciudad)

Su nombres es  juan  y vive en la ciudad de  cordoba


In [6]:
print(f'Su nombre es {nombre} y vive en la ciudad de {ciudad}')

Su nombre es juan y vive en la ciudad de cordoba


Hay 3 formas de hacer este formateo de strings

* El metodo mas antiguo (Python 2) implica el uso del operador de modulo`%`.
* Un metodo mejorado implica el uso del metodo `.format()`.
* El metodo mas nuevo, introducido con Python 3.6, usa lo conocido como *f-strings*.


## Formateo con % (placeholders)
Se usa el operador <code>%</code> para insertar los elementos dentro del string.

Podemos usarlo como <code>%s</code>, <code>%d</code> o <code>%f</code> dependiendo del tipo de dato que queramos insertar.


In [10]:
# %s va a convertir todos los valores que se inserten en strings
print("Mi nombre es %s" %'ivan')

Mi nombre es ivan


Se pueden pasar varios items, solo tenemos que ponerlos entre parentesis separados por coma.

In [11]:
print("Mi nombre es %s y su nombre es %s" %('ivan',nombre))

Mi nombre es ivan y su nombre es juan


In [13]:
print("El numero %s es primo" % 17)

El numero 17 es primo


In [12]:
print("El numero %d es primo" % 17)

El numero 17 es primo


In [15]:
# el operador %d va a convertir los valores numericos en entero
# el operador %f va a convertir los valores numericos en flotante

print('Hoy escribi %s oraciones.' %3.75)
print('Hoy escribi %d oraciones.' %3.75)
print('Hoy escribi %f oraciones.' %3.75)

Hoy escribi 3.75 oraciones.
Hoy escribi 3 oraciones.
Hoy escribi 3.755561 oraciones.


In [ ]:
# esto puede modificarse, al final del notebook hay links para mas info respecto a estos temas

### No dar este tema
Floating point numbers use the format <code>%5.2f</code>. Here, <code>5</code> would be the minimum number of characters the string should contain; these may be padded with whitespace if the entire number does not have this many digits. Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point. Let's see some examples:

In [ ]:
print('Floating point numbers: %5.2f' %(13.144))

Para mas informacion de formateo de strings con placeholders visitar la pagina

https://docs.python.org/3/library/stdtypes.html#old-string-formatting

## Fomateo con el metodo `.format()`
La sintaxis es la siguiente

A better way to format objects into your strings for print statements is with the string `.format()` method. The syntax is:

    'Un string por aqui {} y otro por aqui {}'.format('string_1','string_2')
    
Por ejemplo:

In [16]:
print('Mi nombre es {} y vivo en la ciudad de {}'.format(nombre,ciudad))

Mi nombre es juan y vivo en la ciudad de cordoba


### El metodo .format() tiene varias ventajas sobre el anterior

#### 1. Los objetos a insertar pueden ser llamados por su posicion

In [19]:
print('El {} {} {}'.format('oso','negro','gran'))

El oso negro gran


In [20]:
print('El {2} {0} {1}'.format('oso','negro','gran'))

El gran oso negro


#### 2. Los objetos a insertar pueden ser asignados a variables dentro del .format

In [27]:
print('Primer Objecto: {a}, Segundo Objecto: {b}, Tercer Objecto: {c}'.format(a=1,b='Dos',c=12.3))

Primer Objecto: 1, Segundo Objecto: Dos, Tercer Objecto: 12.3


#### 3. Los objetos pueden ser reutilizados, evitando repeticion.

In [26]:
print('Mi %s es tu %s.' %('casa','casa'))
# vs.
print('Mi {c} es tu {c}.'.format(c='casa'))

Mi casa es tu casa.
Mi casa es tu casa.


### No dar este tema
Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more

Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [20]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

Para mas informacion sobre el metodo `.format()` visitar la pagina 

https://docs.python.org/3/library/string.html#formatstrings

## Formatted String Literals (f-strings)

Este metodo fue introducido con Python 3.6. Ofrece la ventaja de que podemos insertar variables exteriores de manera inmediata.

La sintaxis es:

      f'Inserte el string aqui {"elemento_insertado"}'

Veamos un ejemplo

In [21]:
nombre = 'Valentina'

print(f"Su nombre es {nombre}.")

He said his name is Fred.


## No dar este tema

#### Float formatting follows `"result: {value:{width}.{precision}}"`

Where with the `.format()` method you might see `{value:10.4f}`, with f-strings this can become `{value:{10}.{6}}`


In [23]:
num = 23.45678
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568


Note that with f-strings, *precision* refers to the total number of digits, not just those following the decimal. This fits more closely with scientific notation and statistical analysis. Unfortunately, f-strings do not pad to the right of the decimal, even if precision allows it:

In [24]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:     23.45


If this becomes important, you can always use `.format()` method syntax inside an f-string:

In [25]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:   23.4500


Para mas informacion visitar

https://docs.python.org/3/reference/lexical_analysis.html#f-strings